### ISSUE TITLE
Autosave Featue

### ISSUE THREAD

**Issue by JuBan1** [notepadqq/notepadqq#266]
Looking to implement Autosaving next. Here's what I thought up:

* Autosave creates a file backup of changed files every X minutes. The Y most recent backups will be kept; they're labeled by date and users can select any of them if they want to restore a file from backup. 
* Users can choose the backup interval, number of backups kept, how long backups are kept, and whether to enable or disable autosaves completely.
* Autosave stores the file backups in the cache directory (like sessions) so directories aren't cluttered with .bak files.

But there are a number of things I'm unsure about:

* Should unsaved changes be autosaved? Basically, should be autosave be made from the document currently loaded or the underlying file?

* Likewise, should "orphaned" (new'd documents that weren't opened from a file) documents be autosaved?

* It's difficult to determine when the autosave backups should be deleted. If a user simply closes a file and doesn't ever open it anymore then the autosave files would remain. Best solution I came up with is to put a time-limit on the duration of the autosave files. So they'll be deleted after they are older than Z days.

* The autosave cache could possibly grow quite sizable. An extreme example: user wants to keep 10 last backups of several 1MiB+ size files. That could easily amount to 10s to 100s of MiB depending on how many documents they work on. Maybe add a label to the settings menu of how much space the autosave cache currently uses up? Maybe even an option to limit to cache to X MiB?

Thoughts?

**Comment by Teklad** [comment]
Autosave should definitely operate off the unsaved changes.

I don't think its really necessary to autosave "New" documents that are 
never saved to a file.  The sessions already take care of use cases 
where they want to start back up on the new document.

As for cleaning the autosave... Maybe do a checksum on the saved file vs 
the temporary autosave file... if they match at the time of exit, 
there's no reason to keep both.  Also, clearing the autosave after a 
successful save call would also suffice in these cases.

**Comment by danieleds** [comment]
I'm with @Teklad with this.

The question to ask is: what problem are backup files trying to solve? I believe the big advantage is safety in case of editor crash, kernel panic, asteroids, etc.
If so, clearing the backup files after a successful save plus when the tab is closed, should be sufficient. This is also advisable for privacy reasons.

I agree to keep them in the cache directory.

We should also try to find a minimal set of user-configurable settings so that we keep a compromise between configurability and insanity.

In the weekend I should be able to look at pending pull requests and to discuss more about this :)

**Comment by dagostinelli** [comment]
I'm a little skeptical about what autosave is adding.  I like how the session already works.  

Also, it sounds like there are two major features here (autosave and backup) Concerning backup, I don't much like the idea of .bak files hanging around after I open/close files. It ends up becoming a secondary copy of every text file that I open.   You could probably mitigate this by only back up copying things that I actually change at least once.  

I'd much prefer other features added such as: better handling of indenting with tabs and then aligning with spaces. hex editor / viewer, auto-format json strings, auto-format C code with Allman vs K&R bracing.

**Comment by JuBan1** [comment]
I agree, this is somewhat mixing backups and autosaving. So when I'm looking at it purely from the point of view of wanting to protect a user from unexpected crashes/power outages/etc there'd be two ways that come to mind:

* Nqq could simply save all documents every X minutes/seconds as if the user had pressed CTRL+S. That would be an autosave in the most literal sense but it does change how Nqq behaves and it might conflict with other programs accessing opened files.

* Nqq could basically periodically save the session of each opened window. After a crash these sessions could be re-opened automagically.

**Comment by Teklad** [comment]
Implementing crash protection into the sessions feature wouldn't be a 
bad way to go about this, actually.... and it would keep things cleaner 
than having to implement an extra sub-system for such a small task.

**Comment by dagostinelli** [comment]
@JuBan1 I like  your section option -- periodically save the session (I though it already did that actually, so goes to show) without saving the actual files (so that when it re-launches I see unsaved files as "red").
